### PR TITLE
DROTH-26 Public transport lane to have period of validity to form and Kalpa-API 

### PR DESCRIPTION
**Pull request by MAC-CGI**

### CODE REVIEW

File: UI/src/view/AssetFormFactory.js [hunk 1]
```diff
@@ -551,148 +578,108 @@
...
+      {name: 'integer', fieldType: IntegerField},
+      {name: 'number', fieldType: NumericalField},
+      {name: 'text', fieldType: TextualField},
+      {name: 'checkbox', fieldType: CheckboxField},
+      {name: 'read_only_number', fieldType: ReadOnlyFields},
+      {name: 'read_only_text', fieldType: ReadOnlyFields},
+      {name: 'time_period', fieldType: TimePeriodField}
+  ];
+
+  //TODO change the name to DynamicAssetForm
```
Comment (leirasf) [inline UI/src/view/AssetFormFactory.js]: Do this TODO

File: UI/src/view/AssetFormFactory.js [hunk 2]
```diff
@@ -952,29 +894,121 @@
...
-    var renderLinktoWorkList = function renderLinktoWorkList(layerName) {
-      var textName;
-      switch(layerName) {
-        case "maintenanceRoad":
-          textName = "Tarkistamattomien huoltoteiden lista";
-          break;
-        default:
-          textName = "Vanhentuneiden kohteiden lista";
+    me.isSplitOrSeparatedAllowed = function(){
+      var numberFields = function(sideCode) {
```
Comment (leirasf) [inline UI/src/view/AssetFormFactory.js]: This function don't count the number of fields it should be called number of property values.

File: UI/src/view/AssetFormFactory.js [hunk 3]
```diff
@@ -952,29 +894,121 @@
...
+        return _.flatten(_.map(forms.getFields(sideCode) , function(property) {
+          return property.getPropertyValue().value;
+        })).length;
+      };
+
+      if(_.isEmpty(forms.getAllFields()))
+        return false;
+
+      if (numberFields('a') !== numberFields('b'))
```
Comment (leirasf) [inline UI/src/view/AssetFormFactory.js]: You are not removing duplicated values, before the comparation

File: UI/src/view/AssetFormFactory.js [hunk 4]
```diff
@@ -952,29 +894,121 @@
...
+        return true;
+
+      return _.some(forms.getFields('a'), function(fieldA){
+        var propertyValueA = fieldA.getPropertyValue();
+
+        var propertyValueB = _.head(_.map(_.filter(forms.getFields('b'), function (fieldB) {
```
Comment (leirasf) [inline UI/src/view/AssetFormFactory.js]: First you should add a new function on the dynamicField called getPublicId() then what you want here is
```
var fieldB = _.find(forms.getFields('b'), function(field) { return field.getPublicId() == propertyValueA.publicId; } );
if(!fieldB)
  return true;
return !fieldA.compare(propertyValueA, fieldB.getPropertyValues()));
```

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/dao/MultiValueLinearAssetDao.scala [hunk 5]
```diff
@@ -238,17 +223,33 @@ class MultiValueLinearAssetDao {
...
         } else if (datePropertyValueDoesNotExist(assetId, propertyId)) {
-          insertDateProperty(assetId, propertyId, formatter.parseDateTime(propertyValues.head.value.toString)).execute
+          insertDateProperty(assetId, propertyId, dateFormatter.parseDateTime(propertyValues.head.value.toString)).execute
         } else {
-          updateDateProperty(assetId, propertyId, formatter.parseDateTime(propertyValues.head.value.toString)).execute
+          updateDateProperty(assetId, propertyId, dateFormatter.parseDateTime(propertyValues.head.value.toString)).execute
+        }
+
+      case TimePeriod =>
+        //This type doesn't allow update
```
Comment (MAC-CGI) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/dao/MultiValueLinearAssetDao.scala]: Remove this comment

File: UI/src/view/dynamicAssetForm.js [hunk 6]
```diff
@@ -0,0 +1,1010 @@
...
+        $(event.target).parent().parent().remove();
+        me.setSelectedValue(setValue, getValue);
+      });
+
+      me.element.on('change', '.existing-validity-period .select', function(event) {
+        me.setSelectedValue(setValue, getValue);
+      });
+
+      me.element.on('change', '.new-validity-period select', function(event) {
+        disabled = "";
```
Comment (oskarmalmstrom) [inline UI/src/view/dynamicAssetForm.js]: is this 'disabled' declared somewhere else?

File: UI/src/view/dynamicAssetForm.js [hunk 7]
```diff
@@ -0,0 +1,1010 @@
...
+  var ReadOnlyFields = function(){
+    DynamicField.call(this);
+    var me = this;
+
+    me.editModeRender = function (fieldValue) {
+      var value = _.first(fieldValue, function(values) { return values.value ; });
+      var _value = value ? value.value : '-';
+      return $('' +
+        '<div class="form-group">' +
+        '   <label class="control-label">' + field.label + '</label>' +
```
Comment (oskarmalmstrom) [inline UI/src/view/dynamicAssetForm.js]: field is not declared. should it be fieldValue or something else?

File: digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/IntegrationApi.scala [hunk 8]
```diff
@@ -256,6 +267,32 @@ class IntegrationApi(val massTransitStopService: MassTransitStopService) extends
...
+  def massTransitLanesToApi(typeId: Int, municipalityNumber: Int): Seq[Map[String, Any]] = {
+    def isUnknown(asset:PieceWiseLinearAsset) = asset.id == 0
+
+    val massTransitLanes: Seq[PieceWiseLinearAsset] = multiValueLinearAssetService.getByMunicipality(typeId, municipalityNumber).filterNot(isUnknown)
+
+    massTransitLanes.map { massTransitLane =>
+      Map("id" -> massTransitLane.id,
+        "points" -> massTransitLane.geometry,
+        geometryWKTForLinearAssets(massTransitLane.geometry),
+        "value" -> 1, //check if could be forced
```
[the remark this answers is not in the document]
Comment (MAC-CGI) [reply digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/IntegrationApi.scala]: Maybe "value" field could be removed, Needs to be checked.

File: UI/src/view/dynamicAssetForm.js [hunk 9]
```diff
@@ -0,0 +1,1010 @@
...
+
+  var CheckboxField = function(assetTypeConfiguration, field, isDisabled){
+    DynamicField.call(this, field, isDisabled);
+    var me = this;
+
+    me.editModeRender = function (fieldValue, sideCode, setValue, getValue) {
+      var value = _.first(fieldValue, function(values) { return values.value ; });
+      var _value = value ? value.value : field.defaultValue ? field.defaultValue : "0";
+
+      var checked = !!parseInt(_value) ? 'checked' : '';
```
[the remark this answers is not in the document]
Comment (MAC-CGI) [reply UI/src/view/dynamicAssetForm.js]: is intentional, !! you get a boolean

File: UI/src/view/dynamicAssetForm.js (hunk 6 above)
Comment (MAC-CGI) [reply UI/src/view/dynamicAssetForm.js]: No, it was replaced by me.disabled()

File: UI/src/view/dynamicAssetForm.js (hunk 7 above)
Comment (MAC-CGI) [reply UI/src/view/dynamicAssetForm.js]: yes

File: digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/IntegrationApi.scala [hunk 10]
```diff
@@ -226,14 +236,15 @@ class IntegrationApi(val massTransitStopService: MassTransitStopService) extends
 
   def linearAssetsToApi(typeId: Int, municipalityNumber: Int): Seq[Map[String, Any]] = {
     def isUnknown(asset:PieceWiseLinearAsset) = asset.id == 0
+    //TODO there are two method for getLinearAssetService... should be created only one
```
[the remark this answers is not in the document]
Comment (MAC-CGI) [reply digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/IntegrationApi.scala]: It is a suggestion, could be change in other US

**Pull request merged**